src/mono/mono/utils/atomic.h

```diff
@@ -25,7 +25,236 @@ F/MonoDroid( 1568): shared runtime initialization error: Cannot load library: re
 Apple targets have historically being problematic, xcode 4.6 would miscompile the intrinsic.
 */
 
-#if defined(HOST_WIN32)
+#if !defined(__STDC_NO_ATOMICS__) && !defined(MONO_IGNORE_STDATOMIC)
+
+#include<stdatomic.h>
+
+/* we can't have nice things.  InterlockedCompareExchange and __sync_val_compare_and_swap both
+ * return the previous value of dest.  On the other hand C11 atomics return a boolean success/failure value.
+ *
+ * Emulate the value-returning atomic in terms of the bool one.
+ *
+ * 
+ */
+#define mono_atomic_cas_kernel(VALTYPE,dest,exch,comp) do {	\
```

Comment:
`atomic_compare_exchange` updates `comp` (a pointer!) with the actual value at `dest`, so you don't actually need this hack.

Reply:
You know... I shouldn't write code late at night. I read the cppreference page a dozen times tonight and still somehow missed that one sentence